### ISSUE TITLE
Pulling Mapbox key during front end build process

### ISSUE THREAD

**Issue by mmiddaugh** [department-of-veterans-affairs/va.gov-team#36376]
## Issue Description

- [Mapbox security epic](department-of-veterans-affairs/va.gov-team-sensitive#454)
- We need different API keys for production vs lower environment so developers are not able to access the Production key and so we can monitor traffic. 
- API key needs to come from the parameters store during front end build and be environment specific. If you need more information, please reach out to @darkbushido 

---
## Tasks
- [ ] Front end build process needs to pull an environment-specific Mapbox API key from the parameters store. 

## Acceptance Criteria
- [ ] Current Mapbox production key can be recycled.

**Comment by mmiddaugh** [comment]
@timwright12 any idea when this issue will be prioritized. This blocks our ability to mitigate the Mapbox security vulnerability

**Comment by timwright12** [comment]
@mmiddaugh @darkbushido Can you add some examples of how this is currently being used? We're pretty backed up right now, so if your team knows the implementation please do feel free to open a PR for us to review

**Comment by timwright12** [comment]
Going to mark this ticket as blocked until we can get more information from the submitting team. cc @laineymajor

**Comment by darkbushido** [comment]
I think the tokens are currently stored in the following files

[src/applications/gi/utils/mapboxToken.js](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/applications/gi/utils/mapboxToken.js)
[src/applications/facility-locator/utils/mapboxToken.js](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/applications/facility-locator/utils/mapboxToken.js)

they should be added to the file at build time

**Comment by mmiddaugh** [comment]
@timwright12 apologies for the delay...I've been monitoring the wrong ticket. Thank you for your assistance

**Comment by timwright12** [comment]
@darkbushido do you have docs on how are these are generated and exactly what you all are looking for?

**Comment by mmiddaugh** [comment]
The tokens are generated within Mapbox, if that's what you're asking. [This Access Tokens article might be a good place to start.](https://docs.mapbox.com/help/getting-started/access-tokens/)

**Comment by darkbushido** [comment]
I have updated tokens that were generated by someone who has admin on the account. for right now if we can put the current on in ParamStore I can update them or i can send them to you another way.

Right now we only want to change the non production one so putting the current one in ParamStore would work for now.

**Comment by rmessina1010** [comment]
@darkbushido To clarify scope,  

> Right now we only want to change the non production one so putting the current one in ParamStore would work for now.
is what you are needing from us: to  add a parameter in  AWS  OR  export the current values from the mapboxToken.js files under a different alias?
OR is it to replace the current token with a new one (generated  from the [Mapbox.com dashboard](https://account.mapbox.com/auth/signin/?route-to=%22https://account.mapbox.com/%22), for which i do not have authorization)?

**Comment by darkbushido** [comment]
The current contents of the mapboxToken.js files need to come from the ParamStore. The current value in the mapboxToken.js file can be saved in the ParamStore for now. They do need to be saved under two different entries in the ParamStore one for Production and one for all Non Production Enviroments. 

Once we have moved the current keys into the ParamStore I will replace the Non Production with our test key to make sure its configured correctly. We can they do the same with the Production key assuming it all still works.

- The Keys should not be saved plain text in the repo
- The Keys should come from ParamStore
- The Keys need to be different for Production and Non Production enviroments

Once we have the contents of the mapboxTokens.js files being populated from ParamStore
- We will change the Non Production key in the ParamStore so on the next release we can make sure nothing breaks.
- Once we have fixed any issues with the Non Production key we can then update the production key

**Comment by timwright12** [comment]
@darkbushido I'm not sure we understand what you mean by "ParamStore". This sounds like a very specific ask scoped to a single application. If your team fully understands what you're looking for are you all able to submit a PR and we'll review?

**Comment by rmessina1010** [comment]
Was this what @darkbushido was referring  to: https://docs.aws.amazon.com/systems-manager/latest/userguide/parameter-create-console.html

**Comment by mydesignrocks** [comment]
> The current contents of the mapboxToken.js files need to come from the ParamStore. The current value in the mapboxToken.js file can be saved in the ParamStore for now. They do need to be saved under two different entries in the ParamStore one for Production and one for all Non Production Enviroments.
> 
> Once we have moved the current keys into the ParamStore I will replace the Non Production with our test key to make sure its configured correctly. We can they do the same with the Production key assuming it all still works.
> 
> * The Keys should not be saved plain text in the repo
> * The Keys should come from ParamStore
> * The Keys need to be different for Production and Non Production enviroments
> 
> Once we have the contents of the mapboxTokens.js files being populated from ParamStore
> 
> * We will change the Non Production key in the ParamStore so on the next release we can make sure nothing breaks.
> * Once we have fixed any issues with the Non Production key we can then update the production key

@darkbushido - following up on this ticket now. I should be able to help you with the paramstore updates. Is there a key/value pair that you have in mind to be plugged in as? I do not see a specific one for `mapbox` now, so I am assuming this would be a new entry. You can ping me the key/value details on keybase... my account id is "svalluru2".

**Comment by rmessina1010** [comment]
@darkbushido :
Lance, can you confirm for Sri that what you  need is, once  Sri has created the key/value pairs, as requested  --and I  believe we need one for production and one for non production--, in  AWS we can update  the following:
src/applications/gi/utils/mapboxToken.js
src/applications/facility-locator/utils/mapboxToken.js
to  pull the ENV variables in for you.  I am correct?

**Comment by darkbushido** [comment]
Are there any updates on this?

**Comment by rmessina1010** [comment]
@darkbushido  am huddling with @mydesignrocks  today to coordinate the .env variable name and see if those work with the React app.(will update this comment once that is done)

**Comment by rmessina1010** [comment]
Still working on this. Having issues with webpack/eslint, possibly related to each other.

**Comment by rmessina1010** [comment]
Testing the import (Since this is drawing from Amazon (AWS) it has to be does in an semi-live environment at least)
Our second concern is AWS credentials, in order to access AWS we needed to make sure credentials were being passed  securely. Otherwise this, would hid their tokens, but  reveal  an even more sensitive AWS token. Sri also voiced  concern that other teams might be able to access the token
Theo found that there is an extra dependency (AWS-SDK) that needs to be installed, which may solve the AWS credential issues.

**Comment by rmessina1010** [comment]
**TLDR; This request  may require the retooling of all apps implementing Mapbox and/or mapboxToken; the commit I have made today is for POC/testing and actually breaks all instances of mapbox in our website [read: DO NOT MERGE]**

This is a discovery that happed, tangentially, during the course of addressing the initial issue. 

• Sri determined that In oder to  access the AWS ParamStore, so that we could pull the request variables, we would need add [aws-param-store node plug-in](https://www.npmjs.com/package/aws-param-store) as a dependency; which itself requires the [aws-sdk node plug-in](https://www.npmjs.com/package/aws-sdk) as a sub dependency.

• I was having problems installing the aws-param-store plugin. Theo figured it is possible, if a bit more verbose, to access the ParamStore using only awd-sdk.

• Turns out BOTH plugins work, not by creating an environment,  but by making an async API call to Amazon and thus returning a promise which resolves to the requested data. (It's important to note that javascript doesn't allow for exporting   variables from inside promises or async functions)

• The best we can do is return the promise we get back from Amazon, which will resolve ( contain  the token..inside it).

• However, would mean we would have to retool ALL our apps [which falls outside the scope our team] that implement mapboxToken( they would have to expect a  promise instead of a string, and thus become asynchronous functions themselves).  

• This ALSO means we would need to retool any app that receives any derivative of mapboxToken.

**Comment by mmiddaugh** [comment]
@laineymajor 
@davidconlon and I just discussed this issue. We believe the Facilities team is responsible for the applications which use Mapbox -> Facility Locator, plus map slices for VAMC and Vet Center pages. Can we work together to figure out the tasks to be done and a plan for how to switch over to the new key once the apps have been retooled?

**Comment by rmessina1010** [comment]
@mmiddaugh 


Clarification: the keys (the actual secret word/values) themselves were not  changed.  Only the storage and delivery method were,  as per the original request. We have stored the values  in AWS ParamStore.  Which lead to the issues I was addressing earlier. 

At the moment we have two possible routes we could follow, with  caveats.
* There is the one I delved on in  my previous reply.  I would be happy to coordinate  with someone from the Facilities team (and anyone else that would work with this token) to plan how they would need to handle the new data type. But, full disclosure I anticipate it would be very involved work.

* The other option,  which has proven to be some what experimental, is able to retain  the original string datatype. The caveat  with this approach is that , tho it seems  to be working  properly running locally [mapbox shows!],  it fails/breaks CI tests, specifically--Mocha Tests. I am working with Theo Thursday to isolate the specific fails and reach out  to QA about possible work arounds.

We could also work on both solution concurrently to hedge out bets.

**Comment by rmessina1010** [comment]
For hand-off:
[MapboxToken Conversion Guide Proposal draft](https://vfs.atlassian.net/wiki/spaces/~96033397/pages/2139783260/MapboxToken+Conversion+Guide+Proposal+draft)

**Comment by swirtSJW** [comment]
@timwright12 or @rmessina1010 Do either of you know who owns these keys?

**Comment by rmessina1010** [comment]
The keys are hard coded to each app so I was thinking: @department-of-veterans-affairs/vsa-facilities-frontend and @department-of-veterans-affairs/afs-code-reviewers

**Comment by timwright12** [comment]
@swirtSJW it would be the facilities and/or public websites team. I believe that's over with Agile6 now

**Comment by laineymajor** [comment]
Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VSP---Release-Tools-FE-5fc9325744944e0015ed1861/issues/department-of-veterans-affairs/va.gov-team/36376?planning-poker) with ZenHub @rmessina1010

**Comment by mmiddaugh** [comment]
This security issue is now complete and resolved
